core/src/replay_stage.rs

```diff
@@ -1695,6 +1736,66 @@ impl ReplayStage {
         }
     }
 
+    #[allow(clippy::too_many_arguments)]
+    fn process_repair_threshold_slots(
+        repair_threshold_slots_receiver: &RepairThresholdSlotsReceiver,
+        blockstore: &Blockstore,
+        duplicate_slots_tracker: &mut DuplicateSlotsTracker,
+        repair_threshold_slots: &mut RepairThresholdSlots,
+        epoch_slots_frozen_slots: &mut EpochSlotsFrozenSlots,
+        bank_forks: &RwLock<BankForks>,
+        progress: &ProgressMap,
+        fork_choice: &mut HeaviestSubtreeForkChoice,
+        duplicate_slots_to_repair: &mut DuplicateSlotsToRepair,
+        ancestor_hashes_replay_update_sender: &AncestorHashesReplayUpdateSender,
+        purge_repair_slot_counter: &mut PurgeRepairSlotCounter,
+    ) {
+        let root = bank_forks.read().unwrap().root();
+        for new_repair_threshold_slots in repair_threshold_slots_receiver.try_iter() {
+            for (slot, repair_threshold_hash) in new_repair_threshold_slots {
+                if slot <= root {
+                    continue;
+                }
+                let repairable_hashes = repair_threshold_slots.entry(slot).or_default();
+                if repairable_hashes.contains(&repair_threshold_hash) {
+                    // Already seen
+                    continue;
+                }
+                repairable_hashes.insert(repair_threshold_hash);
+                if repairable_hashes.len() > MAX_REPAIR_THRESHOLD_BLOCKS_PER_SLOT {
+                    error!("We have observed {} > {} blocks with bank hashes {:?} for slot {} that have reached a voted stake > {}.
+                        This is either a critical bug in duplicate block consensus or we are facing an adversary that is more
+                        powerful than our coded assumptions. At this point it is unclear if consensus will be able to converge
+                        and manual intervention might be needed.",
+                        repairable_hashes.len(),
+                        MAX_REPAIR_THRESHOLD_BLOCKS_PER_SLOT,
+                        repairable_hashes,
+                        slot,
+                        REPAIR_THRESHOLD
+                    );
```

Comment:
Should we do anything other than warning? Like not inserting the new hash and break out early?

Reply:
yeah couldn't decide if we should panic or just yell loudly here. 

If we do decide not to panic I think we should still insert the new hash, because that could be the "correct' hash that lets us converge.

Reply:
But someone could manage to send us tons of hashes to make this table big? But it probably takes some effort to make validator OOM.

Reply:
not saying it's the right way to do it but for duplicate consensus v1 we panic once the assumption has been broken:
https://github.com/solana-labs/solana/blob/8a298f1628cb00df1b429acb3fb573ba3491193a/core/src/replay_stage.rs#L1672

so if we wanted to mimic that behavior we would panic here.

Reply:
So I think we set MAX_REPAIR_THRESHOLD_BLOCKS_PER_SLOT for a reason. If we do nothing here other than warning, it seems like this limit is set too low, we can actually handle it but for some reason didn't want to. If incoming slashes exceeding MAX_REPAIR_THRESHOLD_BLOCKS_PER_SLOT will cause any harm downstream, then we should panic.

Reply:
yeah I set it to `5` because that was the maximum that should be achievable by the adversary given the constraints, and the number of columns that would be present in blockstore is based on it. If we're FIFO the columns anyway then we don't need to panic. 

my suggestion would be either:
- Make it panic, but don't merge this PR until legacy shreds are deprecated and quic turbine lets us lock the tree, because then this becomes an easy attack vector. 
- Keep it as error, if we see it pop up in logs we know someone is up to no good.

Reply:
If it's not a lot of trouble, I would prefer a metric, need to know faster if something unexpected is happening, but I'm okay with no panic for now.

Should we even merge this PR until legacy shreds are deprecated?

Reply:
yeah sounds good, i'll just close this pr until legacy shreds are removed.